### PR TITLE
Schema 1.4

### LINKED ISSUE #5509: [Schema] Update schema with Action.Execute properties

### PR DESCRIPTION
**Pull request by dclaux**
## Related Issue
Fixes #5509

## Description
Adds the universal action model definitions to the AC schema
Introduce AC schema version 1.4

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/AdaptiveCards/pull/5317)

### CODE REVIEW

File: schemas/src/AdaptiveCard.json [hunk 1]
```diff
@@ -4,13 +4,18 @@
 	"properties": {
 		"version": {
 			"type": "string",
-		"description": "Schema version that this card requires. If a client is **lower** than this version, the `fallbackText` will be rendered. NOTE: Version is not required for cards within an `Action.ShowCard`. However, it *is* required for the top-level card.",
+			"description": "Schema version that this card requires. If a client is **lower** than this version, the `fallbackText` will be rendered. NOTE: Version is not required for cards within an `Action.ShowCard`. However, it *is* required for the top-level card.",
```
Comment (paulcam206) [inline schemas/src/AdaptiveCard.json:7]: thanks for fixing this! :)

File: schemas/src/actions/Action.Execute.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
+{
+	"$schema": "https://raw.githubusercontent.com/microsoft/AdaptiveCards/6f39aedce45864ae1067ed44a5551dc973790bb5/source/nodejs/typed-schema/schema/lib/Type.json",
+	"extends": "Action, ISelectAction",
+	"description": "Gathers input fields, merges with optional data field, and sends an event to the client. Clients process the event by sending an Invoke activity of type adaptiveCard/action to the target Bot. The inputs that are gathered are those on the current card, and in the case of a show card those on any parent cards. See https://docs.microsoft.com/en-us/adaptive-cards/authoring-cards/input-validation for more details.",
+    "version": "1.4",
```
Comment (paulcam206) [inline schemas/src/actions/Action.Execute.json:5]: nitpick: whitespace

File: schemas/src/actions/Action.Execute.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
+{
+	"$schema": "https://raw.githubusercontent.com/microsoft/AdaptiveCards/6f39aedce45864ae1067ed44a5551dc973790bb5/source/nodejs/typed-schema/schema/lib/Type.json",
+	"extends": "Action, ISelectAction",
```
Comment (paulcam206) [inline schemas/src/actions/Action.Execute.json:3]: just to be sure -- we want `Action.Execute` to be a `selectAction` target, yes?

**Review by paulcam206 — APPROVED** [review]

File: schemas/src/actions/Action.Execute.json (hunk 3 above)
Comment (dclaux) [reply schemas/src/actions/Action.Execute.json:3]: Yes

File: schemas/src/AdaptiveCard.json [hunk 4]
```diff
@@ -4,13 +4,23 @@
...
 			]
 		},
+		"refresh": {
+			"type": "Refresh",
+			"description": "Defines how the card can be refreshed by making a request to the target Bot.",
+			"version": "1.4"
+		},
+		"authentication": {
+			"type": "Authentication",
+			"description": "Defines authentication information to enable on-behalf-f single sign on or just-in-time OAuth.",
```
Comment (paulcam206) [inline schemas/src/AdaptiveCard.json:21]:
```suggestion
			"description": "Defines authentication information to enable on-behalf-of single sign on or just-in-time OAuth.",
```

File: schemas/src/shared/AuthCardButton.json [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+            "type": "string",
+            "description": "The caption of the button."
+        },
+        "image": {
+            "type": "string",
+            "description": "A URL to an image to display alongside the button's caption."
+        },
+        "value": {
+            "type": "string",
+            "description": "The value associated eith the button. The meaning of value depends on the button's type.",
```
Comment (paulcam206) [inline schemas/src/shared/AuthCardButton.json:21]:
```suggestion
            "description": "The value associated with the button. The meaning of value depends on the button's type.",
```

File: schemas/src/shared/TokenExchangeResource.json [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    "version": "1.4",
+    "properties": {
+        "id": {
+            "type": "string",
+            "description": "The unique identified of this token exchange instance.",
+            "required": true
+        },
+        "uri": {
+            "type": "string",
+            "description": "An application ID or resource identifier with which to exchange a token on behalf of. This property is is identity provider- and application-specific.",
```
Comment (paulcam206) [inline schemas/src/shared/TokenExchangeResource.json:13]:
```suggestion
            "description": "An application ID or resource identifier with which to exchange a token on behalf of. This property is identity provider- and application-specific.",
```

**Review by paulcam206 — APPROVED** [review]
just found a couple typos

File: schemas/src/shared/Authentication.json [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    "description": "Defines authentication information associated with a card. This maps to the OAuthCard type defined by the Bot Framework (https://github.com/microsoft/botframework-obi/blob/jeffders/adaptiveCards2/protocols/botframework-activity/botframework-cards.md#oauth-card)",
+    "version": "1.4",
+    "properties": {
+        "text": {
+            "type": "string",
+            "description": "Text that can be displayed to the end user when prompting them to authenticate."
+        },
+        "connectionName": {
+            "type": "string",
+            "description": "The identifier for registered OAuth connenection setting information."
```
[the remark this answers is not in the document]
Comment (dclaux) [reply schemas/src/shared/Authentication.json]: Fixed

File: schemas/src/AdaptiveCard.json (hunk 4 above)
Comment (dclaux) [reply schemas/src/AdaptiveCard.json]: Fixed

File: schemas/src/shared/TokenExchangeResource.json (hunk 6 above)
Comment (dclaux) [reply schemas/src/shared/TokenExchangeResource.json]: Fixed

File: schemas/src/shared/AuthCardButton.json (hunk 5 above)
Comment (dclaux) [reply schemas/src/shared/AuthCardButton.json]: Fixed

File: source/nodejs/adaptivecards-site/schema-explorer-toc.yml [hunk 8]
```diff
@@ -21,6 +21,7 @@
   - Action.ShowCard
   - Action.ToggleVisibility
   - TargetElement
+  - Action.Execute
```
Comment (matthidinger) [inline source/nodejs/adaptivecards-site/schema-explorer-toc.yml:24]: We should also include the new `Refresh` and `Authentication` types at the bottom of this file, under `Types`, so users can view the schema for those

**Review by matthidinger — CHANGES_REQUESTED** [review]
Couple things:
1. We should include the new types like Refresh and Authentication in the TOC so users can nagivage to them
2. Can we include some example payloads? At least the minimum ones such as an Action.Execute example, but additional ones for Refresh would be nice. They go in a new folder, like below one but with the 1.4 version. When added, the samples show up with a "Try it yourself" button, e.g., here https://adaptivecardsci.z5.web.core.windows.net/pr/5317/explorer/Action.Execute.html 

Reference: 
https://github.com/microsoft/AdaptiveCards/tree/main/samples/v1.3/Elements

File: schemas/src/actions/Action.Execute.json [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+	"description": "Gathers input fields, merges with optional data field, and sends an event to the client. Clients process the event by sending an Invoke activity of type adaptiveCard/action to the target Bot. The inputs that are gathered are those on the current card, and in the case of a show card those on any parent cards. See https://docs.microsoft.com/en-us/adaptive-cards/authoring-cards/input-validation for more details.",
+    "version": "1.4",
+	"properties": {
+		"verb": {
+			"type": "string",
+			"description": "The developer-defined verb associated with this action."
```
Comment (golddove) [inline schemas/src/actions/Action.Execute.json:9]: Perhaps "card author" to clarify developer

File: schemas/src/shared/AuthCardButton.json [hunk 10]
```diff
@@ -0,0 +1,25 @@
+{
+    "$schema": "https://raw.githubusercontent.com/microsoft/AdaptiveCards/6f39aedce45864ae1067ed44a5551dc973790bb5/source/nodejs/typed-schema/schema/lib/Type.json",
+    "description": "Defines a button as displayed when prompting a user to authenticate. This maps to the cardAction type defined by the Bot Framework (https://github.com/microsoft/botframework-obi/blob/jeffders/adaptiveCards2/protocols/botframework-activity/botframework-activity.md#card-action).",
+    "version": "1.4",
+    "properties": {
+        "type": {
+            "type": "string",
```
Comment (golddove) [inline schemas/src/shared/AuthCardButton.json:7]: Seems like there are other possible "types", not just "signin", that we may use in the future? If so, is AuthCardButton too specific of a name for this schema?

**Review by golddove — APPROVED** [review]

File: schemas/src/shared/AuthCardButton.json (hunk 10 above)
Comment (dclaux) [reply schemas/src/shared/AuthCardButton.json:7]: It maps directly to the type names defined in the existing Bot protocol. This is a case of reuse of what already existed before Universal Actions, to maximize code reuse.

**Review by golddove — APPROVED** [review]

File: source/nodejs/adaptivecards-site/schema-explorer-toc.yml (hunk 8 above)
Comment (dclaux) [reply source/nodejs/adaptivecards-site/schema-explorer-toc.yml:24]: That's a separate thing though not? This PR only introduces the 1.4 schema.

**Review by jwoo-msft — DISMISSED** [review]

**Review by matthidinger — APPROVED** [review]

**Pull request merged**